x-pack/plugins/elastic_assistant/server/routes/post_actions_connector_execute.ts

```diff
@@ -8,139 +8,286 @@
+              role: c.role,
+              content: c.content,
+            }));
+
+            if (conversation == null) {
+              return response.notFound({
+                body: `conversation id: "${conversationId}" not found`,
+              });
+            }
+
+            if (request.body.replacements) {
+              await dataClient?.updateConversation({
+                conversationUpdateProps: {
+                  id: conversationId,
+                  replacements: request.body.replacements,
+                },
+              });
+            }
+
+            const dateTimeString = new Date().toISOString();
+
+            const appendMessageFuncs = request.body.params.subActionParams.messages.map(
+              (userMessage) => async () => {
+                if (conversation != null) {
+                  const res = await dataClient?.appendConversationMessages({
+                    existingConversation: conversation,
+                    messages: request.body.params.subActionParams.messages.map((m) => ({
+                      content: getMessageContentWithoutReplacements({
+                        messageContent: userMessage.content,
+                        replacements: (request.body.replacements ?? {}) as Record<string, string>,
+                      }),
+                      role: m.role,
+                      timestamp: dateTimeString,
+                    })),
+                  });
+
+                  if (res == null) {
+                    return response.badRequest({
+                      body: `conversation id: "${conversationId}" not updated`,
+                    });
+                  }
+                }
+              }
+            );
+
+            await Promise.all(appendMessageFuncs.map((appendMessageFunc) => appendMessageFunc()));
+
+            const updatedConversation = await dataClient?.getConversation({
+              id: conversationId,
+              authenticatedUser,
+            });
+
+            if (updatedConversation == null) {
+              return response.notFound({
+                body: `conversation id: "${conversationId}" not found`,
+              });
+            }
+
+            onMessageSent = async (
+              content: string,
+              traceData: Message['traceData'] = {}
+            ): Promise<void> => {
+              if (updatedConversation) {
+                await dataClient?.appendConversationMessages({
+                  existingConversation: updatedConversation,
+                  messages: [
+                    getMessageFromRawResponse({
+                      rawContent: getMessageContentWithoutReplacements({
+                        messageContent: content,
+                        replacements: (request.body.replacements ?? {}) as Record<string, string>,
+                      }),
+                      traceData,
+                    }),
+                  ],
+                });
+              }
+            };
+          }
```

Comment:
Consider extracting these new statements into one or more new functions, because it's hard to keep track of what happens here and then reason about what happens in the existing code below (i.e. the call to `executeAction()`)